stream-chat-android-ui-common/src/main/kotlin/io/getstream/chat/android/ui/common/feature/channel/info/ChannelInfoController.kt

```diff
@@ -0,0 +1,420 @@
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.SharedFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asSharedFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.flow.flatMapLatest
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+
+/**
+ * Controller responsible for managing the state and events related to channel information.
+ *
+ * It provides functionality to observe channel data, members, and perform
+ * various channel-related actions such as renaming, muting, hiding, leaving, and deleting the channel.
+ * It also handles state updates and emits events for UI consumption.
+ *
+ * @param cid The unique identifier of the channel.
+ * @param scope The [CoroutineScope] used for launching coroutines.
+ * @param chatClient The [ChatClient] instance used for interacting with the chat API.
+ * @param channelState A [Flow] representing the state of the channel.
+ * @param channelClient The [ChannelClient] instance for performing channel-specific operations.
+ */
+@InternalStreamChatApi
+public class ChannelInfoController(
+    cid: String,
+    private val scope: CoroutineScope,
+    private val chatClient: ChatClient = ChatClient.instance(),
+    channelState: Flow<ChannelState> = chatClient
+        .watchChannelAsState(cid = cid, messageLimit = 0, coroutineScope = scope)
+        .filterNotNull(),
+    private val channelClient: ChannelClient = chatClient.channel(cid),
+) {
+    private val logger by taggedLogger("Chat:ChannelInfoController")
+
+    private val _state = MutableStateFlow<ChannelInfoState>(ChannelInfoState.Loading)
+
+    /**
+     * A [StateFlow] representing the current state of the channel info.
+     */
+    public val state: StateFlow<ChannelInfoState> = _state.asStateFlow()
+
+    private val _events = MutableSharedFlow<ChannelInfoEvent>(extraBufferCapacity = 1)
+
+    /**
+     * A [SharedFlow] that emits one-time events related to channel info, such as errors or success events.
+     */
+    public val events: SharedFlow<ChannelInfoEvent> = _events.asSharedFlow()
+
+    init {
+        @Suppress("OPT_IN_USAGE")
+        channelState
+            .flatMapLatest { channel ->
+                logger.d { "[onChannelState]" }
+                combine(
+                    channel.channelData.onEach { logger.d { "[onChannelData] name: ${it.name}" } },
+                    channel.members.onEach { logger.d { "[onMembers] size: ${it.size}" } },
+                    channel.muted.onEach { logger.d { "[onMuted] $it" } },
+                    channel.hidden.onEach { logger.d { "[onHidden] $it" } },
```

Comment:
Do we need this log on every update?

Reply:
It was pretty handy when debugging. That's why I used `logger.d`

Reply:
I can review them after have it fully integrated with the view models and the UI 🙏🏻